### PR TITLE
Alternative / simplified serialization API

### PR DESCRIPTION
**Pull request by achoum**
As discussed in https://github.com/google/temporian/pull/9#issuecomment-1421051836

### CODE REVIEW

File: temporian/core/serialize.py [hunk 1]
```diff
@@ -29,7 +29,10 @@
 
 
 def save(
-    inputs: Dict[str, Event], outputs: Dict[str, Event], path: str
+    inputs: Optional[processor.MultipleEventArg],
+    outputs: processor.MultipleEventArg,
+    path: str,
+    infer_inputs: bool = False,
```
Comment (ianspektor) [inline temporian/core/serialize.py:35]: Propose defaulting to True (at least when user doesn't specify `inputs`)

File: temporian/core/serialize.py [hunk 2]
```diff
@@ -49,21 +52,38 @@ def save(
...
 
     # TODO: Add support for compressed / binary serialization.
 
-    p = processor.infer_processor(inputs=inputs, outputs=outputs)
+    if infer_inputs:
+        if inputs is not None:
+            raise ValueError("If infer_inputs=True, inputs should be None")
+    else:
+        if inputs is None:
+            raise ValueError("If infer_inputs=False, inputs should be None")
```
Comment (ianspektor) [inline temporian/core/serialize.py:65]: `inputs shouldn't be None`

File: temporian/test/api_test.py [hunk 3]
```diff
@@ -71,6 +71,79 @@ def test_serialization(self):
         self.assertSetEqual(set(inputs.keys()), {"a"})
         self.assertSetEqual(set(outputs.keys()), {"b"})
 
+    def test_serialization_v2(self):
```
Comment (ianspektor) [inline temporian/test/api_test.py:74]: Nit: some more descriptive names or docstrings in these tests would be nice 😅

File: temporian/core/processor.py [hunk 4]
```diff
@@ -147,15 +181,25 @@ def infer_processor(
...
+    # List the events containing each possible feature:
+    # Used for the automatic input inference.
+    features_to_src_event: Dict[Feature, List[Event]] = defaultdict(lambda: [])
+
     pending_features: Set[Feature] = set()
     for output_event in outputs.values():
         pending_features.update(output_event.features())
         p.add_sampling(output_event.sampling())
+        for feature in output_event.features():
+            features_to_src_event[feature].append(pending_features)
```
[the remark this answers is not in the document]
Comment (achoum) [reply temporian/core/processor.py]: Good catch.

*Fix + add unit test for this case.

File: temporian/core/test/processor_test.py [hunk 5]
```diff
@@ -63,6 +63,55 @@ def test_dependency_missing_input(self):
...
+    def test_automatic_input_missing_name(self):
+        i1 = utils.create_input_event()
+        o2 = utils.OpI1O1(i1)
+
+        with self.assertRaisesRegex(ValueError, "Infered input without a name"):
+            processor.infer_processor(
+                None,
+                {"io_output_1": o2.outputs()["output"]},
+                infer_inputs=True,
+            )
```
[the remark this answers is not in the document]
Comment (achoum) [reply temporian/core/test/processor_test.py]: Unit have been rewritten / improved.

File: temporian/core/serialize.py (hunk 1 above)
Comment (achoum) [reply temporian/core/serialize.py]: *Remove infer_inputs arguments. When inputs=None, the input is inferred automatically.

File: temporian/core/serialize.py (hunk 2 above)
Comment (achoum) [reply temporian/core/serialize.py]: *Removed

File: temporian/test/api_test.py (hunk 3 above)
Comment (achoum) [reply temporian/test/api_test.py]: Done :)

**Review by ianspektor — APPROVED** [review]
Love the simplification of `infer_processor`'s logic after removing the `infer_inputs` arg! Thanks for the changes :)

**Comment by achoum** [comment]
Thanks

**Pull request merged**